crates/sats/src/typespace.rs

```diff
@@ -86,6 +102,73 @@ impl Typespace {
     pub const fn with_type<'a, T: ?Sized>(&'a self, ty: &'a T) -> WithTypespace<'a, T> {
         WithTypespace::new(self, ty)
     }
+
+    /// Inlines all type references in `ty` recursively using the current typeset.
+    pub fn inline_typerefs_in_type(&mut self, ty: &mut AlgebraicType) -> Result<(), TypeRefError> {
+        match ty {
+            AlgebraicType::Sum(sum_ty) => {
+                for variant in &mut sum_ty.variants {
+                    self.inline_typerefs_in_type(&mut variant.algebraic_type)?;
+                }
+            }
+            AlgebraicType::Product(product_ty) => {
+                for element in &mut product_ty.elements {
+                    self.inline_typerefs_in_type(&mut element.algebraic_type)?;
+                }
+            }
+            AlgebraicType::Builtin(BuiltinType::Array(array_ty)) => {
+                self.inline_typerefs_in_type(&mut array_ty.elem_ty)?;
+            }
+            AlgebraicType::Builtin(BuiltinType::Map(map_type)) => {
+                self.inline_typerefs_in_type(&mut map_type.key_ty)?;
+                self.inline_typerefs_in_type(&mut map_type.ty)?;
+            }
+            AlgebraicType::Ref(r) => {
+                // Lazily resolve any nested references first.
+                let resolved_ty = self.inline_typerefs_in_ref(*r)?;
+                // Now we can clone the fully-resolved type.
+                *ty = resolved_ty.clone();
+            }
+            _ => {}
+        }
+        Ok(())
+    }
+
+    /// Inlines all nested references behind the current [`AlgebraicTypeRef`] recursively using the current typeset.
+    ///
+    /// Returns the fully-resolved type or an error if the type reference is invalid or self-referential.
+    fn inline_typerefs_in_ref(&mut self, r: AlgebraicTypeRef) -> Result<&AlgebraicType, TypeRefError> {
+        let resolved_ty = match self.get_mut(r) {
+            None => return Err(TypeRefError::InvalidTypeRef(r)),
+            Some(AlgebraicType::Ref(_)) => return Err(TypeRefError::RecursiveTypeRef(r)),
```

Comment:
Should this be checking that the `Ref` you extracted is equal to `r`? Is there ever a case where we'll have two non-equal refs `A` and `B` such that `A` refers to `B`?

Reply:
```suggestion
            // If `r` refers to itself, that means an enclosing call to `inline_typerefs_in_ref(r)` 
            // swapped its definition out,
            // i.e. the type referred to by `r` is recursive.
            // Our database can't handle recursive types, so return an error.
            // TODO: support recursive types in the future.
            Some(AlgebraicType::Ref(_)) => return Err(TypeRefError::RecursiveTypeRef(r)),
```

Reply:
Added a variation of your comment that explains any of the parent calls might've swapped it out.

Reply:
No, I mean, could we have a non-recursive typespace like:

```
0 -> 1
1 -> U32
```